src/main/java/com/datalogics/pdf/security/LunaHsmManager.java

```diff
@@ -0,0 +1,180 @@
+    public static final String PROVIDER_NAME = "LunaProvider";
+
+    /**
+     * Default no-arg constructor.
+     *
+     * @throws IOException an I/O operation failed or was interrupted
+     */
+    protected LunaHsmManager() throws IOException {
+        super();
+        slotManager = LunaSlotManager.getInstance();
+        lunaKeyStore = getLunaKeyStore();
+        initializeProvider();
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see com.datalogics.pdf.hsm.samples.HsmManager#hsmLogin(java.lang.String, java.lang.String)
+     */
+    @Override
+    public boolean hsmLogin(final HsmLoginParms parms) {
+        if (!(parms instanceof LunaHsmLoginParms)) {
+            throw new IllegalArgumentException("Must pass a instanceof LunaHsmLoginParms "
+                                               + "to hsmLogin for LunaHsmManager.");
+        }
+
+        final String password = parms.getPassword();
+        // Check for non-null, non zero length password
+        if (password == null) {
+            throw new IllegalArgumentException("Password must not be null");
+        } else if (password.length() <= 0) {
+            throw new IllegalArgumentException("Password must not be zero length");
+        }
+
+        final String tokenLabel = ((LunaHsmLoginParms) parms).getTokenLabel();
+        try {
+            if (tokenLabel == null) {
+                slotManager.login(password);
+            } else {
+                slotManager.login(tokenLabel, password);
+            }
+        } catch (LunaException | LunaCryptokiException e) {
+            throw new IllegalArgumentException("Error while logging into the Luna HSM" + e);
+        }
+
+        isLoggedIn = slotManager.isLoggedIn();
+        loadKeyStore();
+        return isLoggedIn;
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see com.datalogics.pdf.hsm.samples.HsmManager#hsmLogout()
+     */
+    @Override
+    public void hsmLogout() {
+        /*
+         * When you are done using the Luna HSM, it is customary to log out of the HSM to prevent unauthorized access at
+         * a later point in your application.
+         *
+         */
+        if (isLoggedIn) {
+            slotManager.logout();
+            isLoggedIn = false;
+        }
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see com.datalogics.pdf.security.HsmManager#isLoggedIn()
+     */
+    @Override
+    public boolean isLoggedIn() {
+        return isLoggedIn;
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see com.datalogics.pdf.security.HsmManager#getCertificateChain(java.lang.String)
+     */
+    @Override
+    public X509Certificate[] getCertificateChain(final String certLabel) {
+        X509Certificate cert;
+        try {
+            cert = (X509Certificate) lunaKeyStore.getCertificate(certLabel);
+        } catch (final KeyStoreException e) {
+            throw new SecurityException("Exception while obtaining certificate chain for LunaSa: ", e);
```

Comment:
- I think all of these getter methods should explicitly check if we are logged in before proceeding, and throw a SecurityException if not. That's likely what happens anyway, but we could return a more detailed message.)
- We should standardize on a spelling for 'Luna SA'